Microservices Example

benstopford opened the pull request:
A small ecosystem of services that provides Orders Functionality over a CQRS-styled interface.  

**OrdersService**
- Accepts orders via REST
- Allows orders to be Queried, blocking if they are not present. (Uses non blocking IO and Queryable state).

**InventoryService / FraudService / OrderDetailsService** 
- Provide different validations of the order, run in parallel. 
- First two demo the KStreams API, on the transactional producer and consumer. 

**Validations Aggregator Service** 
- Aggregates validations together and updates the corresponding order

**TODO** 
- Demo embedded KSQL around the input of Inventory (which can be done without Avro support)
- Add multi-node Kafka support to the test framework. 
- Add tests that use KStreamTestDriver to demonstrate how that works.

dguy commented on src/main/java/io/confluent/examples/streams/microservices/OrderDetailsService.java:
```diff
@@ -0,0 +1,154 @@
+        if (order.getCustomerId() == null) return false;
+        if (order.getQuantity() < 0) return false;
+        if (order.getPrice() < 0) return false;
+        if (order.getProduct() == null) return false;
```
`return order.getProduct() != null` ?

benstopford commented:
Thanks @dguy. On the orderid being versioned, it could be either, but the advantage of having the version on the OrderId is there is no way for the processing of one version of an order to affect another. So say I get two order updates in quick succession

benstopford commented on src/main/java/io/confluent/examples/streams/microservices/OrderDetailsService.java:
```diff
@@ -0,0 +1,154 @@
+ * including the integration with Kafka's Exactly Once feature (Transactions) to demonstrate
+ * this other style of building event driven services.
+ */
+public class OrderDetailsService implements Service {
```
just to show that you can really, but maybe that's sending the wrong message.

benstopford commented:
@dguy I've addressed all your comments. Thank you for them. I'm going to leave the other TODO items in the description above for a follow up PR as I'd like to get this one merged. You ok to lgtm?
